File: tests/oracle/funcproto_effect_model_1204.cc
/*
 * PROTO-EFFECT-MODEL-0001: locked Ghidra 12.0.4 FuncProto model/effect oracle.
 */
#include "bfd_arch.hh"
#include "fspec.hh"
#include "libdecomp.hh"
#include "marshal.hh"

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;

struct Probe {
  const char *label;
  const char *space;
  uintb offset;
  int4 size;
};

const Probe probes[] = {
  { "unique_always", "unique", 0x777, 4 },
  { "same_offset_const", "const", 0x10, 8 },
  { "whole_ram_space", "ram", 0x777, 32 },
  { "register_exact", "register", 0x10, 8 },
  { "register_contained", "register", 0x12, 2 },
  { "register_partial_left", "register", 0x0e, 4 },
  { "register_partial_right", "register", 0x16, 4 },
  { "register_killed", "register", 0x20, 8 },
  { "return_address", "register", 0x30, 8 },
  { "same_offset_stack", "stack", 0x10, 8 },
  { "same_offset_register", "register", 0x10, 8 },
  { "before_first_register", "register", 0x08, 8 }
};

const char *effectName(uint4 effect);

void writeEffectiveRecords(ostream &out,const FuncProto &proto)
{
  out << '[';
  bool first = true;
  for(vector<EffectRecord>::const_iterator iter=proto.effectBegin();
      iter!=proto.effectEnd();++iter) {
    if (!first) out << ',';
    first = false;
    Address address = iter->getAddress();
    out << "{\"space\":\"" << address.getSpace()->getName()
        << "\",\"offset\":" << address.getOffset()
        << ",\"size\":" << iter->getSize()
        << ",\"effect\":\"" << effectName(iter->getType()) << "\"}";
  }
  out << ']';
}

const char *effectName(uint4 effect)
{
  switch(effect) {
  case EffectRecord::unaffected: return "unaffected";
  case EffectRecord::killedbycall: return "killedbycall";
  case EffectRecord::return_address: return "return_address";
  case EffectRecord::unknown_effect: return "unknown_effect";
  default: return "invalid";
  }
}

void decodeModel(Architecture &architecture,ProtoModel &model,const string &xml)
{
  istringstream stream(xml);
  XmlDecode decoder(&architecture);
  decoder.ingestStream(stream);
  model.decode(decoder);
}

void decodeProto(Architecture &architecture,ProtoModel *model,Datatype *voidType,
                 FuncProto &proto,const string &xml)
{
  proto.setInternal(model,voidType);
  istringstream stream(xml);
  XmlDecode decoder(&architecture);
  decoder.ingestStream(stream);
  proto.decode(decoder,&architecture);
}

void writeEffects(ostream &out,const FuncProto &proto,const Architecture &architecture)
{
  out << '[';
  for(size_t i=0;i<sizeof(probes)/sizeof(probes[0]);++i) {
    if (i != 0) out << ',';
    AddrSpace *space = architecture.getSpaceByName(probes[i].space);
    if (space == (AddrSpace *)0)
      throw std::runtime_error(string("missing address space: ") + probes[i].space);
    out << "{\"label\":\"" << probes[i].label
        << "\",\"space\":\"" << probes[i].space
        << "\",\"offset\":" << probes[i].offset
        << ",\"size\":" << probes[i].size
        << ",\"effect\":\""
        << effectName(proto.hasEffect(Address(space,probes[i].offset),probes[i].size))
        << "\"}";
  }
  out << ']';
}

void runFixture(const string &specDirectory,const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  BfdArchitecture architecture(binary,"default",&std::cerr);
  DocumentStorage documentStorage;
  architecture.init(documentStorage);

  Datatype *voidType = architecture.types->getTypeVoid();
  if (voidType == (Datatype *)0)
    throw std::runtime_error("architecture has no void type");

  const string modelXml =
    "<prototype name=\"fixture_model\" extrapop=\"16\" hasthis=\"true\" constructor=\"true\">"
    "<input/>"
    "<output killedbycall=\"true\"/>"
    "<unaffected><addr space=\"ram\" offset=\"0\" size=\"0\"/>"
    "<addr space=\"register\" offset=\"16\" size=\"8\"/></unaffected>"
    "<killedbycall><addr space=\"const\" offset=\"16\" size=\"8\"/>"
    "<addr space=\"register\" offset=\"32\" size=\"8\"/>"
    "<addr space=\"stack\" offset=\"16\" size=\"8\"/></killedbycall>"
    "<returnaddress><addr space=\"register\" offset=\"48\" size=\"8\"/></returnaddress>"
    "</prototype>";
  const string unknownModelXml =
    "<prototype name=\"fixture_unknown\" extrapop=\"unknown\">"
    "<input/><output/><returnaddress>"
    "<addr space=\"register\" offset=\"64\" size=\"8\"/>"
    "</returnaddress></prototype>";
  const string replacementModelXml =
    "<prototype name=\"fixture_replacement\" extrapop=\"24\">"
    "<input/><output/></prototype>";

  ProtoModel *model = new ProtoModel(&architecture);
  decodeModel(architecture,*model,modelXml);
  architecture.protoModels[model->getName()] = model;

  ProtoModel independentModel(&architecture);
  decodeModel(architecture,independentModel,modelXml);
  ProtoModel unknownModel(&architecture);
  decodeModel(architecture,unknownModel,unknownModelXml);
  ProtoModel replacementModel(&architecture);
  decodeModel(architecture,replacementModel,replacementModelXml);

  FuncProto modelProto;
  modelProto.setInternal(model,voidType);
  FuncProto modelCopy;
  modelCopy.copy(modelProto);
  FuncProto independentProto;
  independentProto.setInternal(&independentModel,voidType);

  const string overrideAXml =
    "<prototype model=\"fixture_model\" extrapop=\"16\">"
    "<returnsym><addr/><void/></returnsym>"
    "<killedbycall><addr space=\"register\" offset=\"16\" size=\"8\"/></killedbycall>"
    "</prototype>";
  const string overrideBXml =
    "<prototype model=\"fixture_model\" extrapop=\"16\">"
    "<returnsym><addr/><void/></returnsym>"
    "<unaffected><addr space=\"register\" offset=\"48\" size=\"8\"/></unaffected>"
    "</prototype>";
  FuncProto overrideA;
  decodeProto(architecture,model,voidType,overrideA,overrideAXml);
  FuncProto overrideACopy;
  overrideACopy.copy(overrideA);
  FuncProto overrideB;
  decodeProto(architecture,model,voidType,overrideB,overrideBXml);
  overrideA.copy(overrideB);

  AddrSpace *registerSpace = architecture.getSpaceByName("register");
  VarnodeData lookupStorage;
  lookupStorage.space = registerSpace;
  lookupStorage.offset = 0x10;
  lookupStorage.size = 8;
  vector<EffectRecord> lookupRecords;
  lookupRecords.push_back(EffectRecord(lookupStorage,EffectRecord::unaffected));
  int4 beforeFirstOverlap = ProtoModel::lookupRecord(
    lookupRecords,1,Address(registerSpace,0x08),16);
  int4 beforeFirstDisjoint = ProtoModel::lookupRecord(
    lookupRecords,1,Address(registerSpace,0x00),8);

  FuncProto sticky;
  sticky.setInternal(model,voidType);
  int4 initialExtraPop = sticky.getExtraPop();
  bool initialThis = sticky.hasThisPointer();
  bool initialConstructor = sticky.isConstructor();
  bool initialAutoKilled = sticky.isAutoKilledByCall();
  sticky.setModel(&unknownModel);
  int4 switchedExtraPop = sticky.getExtraPop();
  bool switchedThis = sticky.hasThisPointer();
  bool switchedConstructor = sticky.isConstructor();
  bool switchedAutoKilled = sticky.isAutoKilledByCall();
  sticky.setModel(&replacementModel);
  int4 replacedExtraPop = sticky.getExtraPop();
  bool replacedThis = sticky.hasThisPointer();
  bool replacedConstructor = sticky.isConstructor();
  bool replacedAutoKilled = sticky.isAutoKilledByCall();
  sticky.setModel((ProtoModel *)0);
  ostringstream nullModelPrint;
  sticky.printRaw("fixture",nullModelPrint);
  bool nullPrintHasNoModel =
    nullModelPrint.str().compare(0,11,"(no model) ") == 0;

  FuncProto outputLocked;
  outputLocked.setInternal((ProtoModel *)0,voidType);
  bool outputLockBefore = outputLocked.isAutoKilledByCall();
  outputLocked.setOutputLock(true);
  bool outputLockAfter = outputLocked.isAutoKilledByCall();

  FuncProto copyHintSource;
  copyHintSource.setInternal((ProtoModel *)0,voidType);
  copyHintSource.setReturnBytesConsumed(7);
  FuncProto copyHintDestination;
  copyHintDestination.setInternal((ProtoModel *)0,voidType);
  copyHintDestination.setReturnBytesConsumed(3);
  copyHintDestination.copy(copyHintSource);

  std::cout << "{\"schema\":1,\"fixture\":\"PROTO-EFFECT-MODEL-0001\""
            << ",\"model_effects\":";
  writeEffects(std::cout,modelProto,architecture);
  std::cout << ",\"effective_model_records\":";
  writeEffectiveRecords(std::cout,modelProto);
  std::cout << ",\"model_identity\":{\"copy\":"
            << modelCopy.hasMatchingModel(model)
            << ",\"independent_same_definition\":"
            << independentProto.hasMatchingModel(model)
            << "},\"override_a_copy\":";
  writeEffects(std::cout,overrideACopy,architecture);
  std::cout << ",\"effective_override_records\":";
  writeEffectiveRecords(std::cout,overrideACopy);
  std::cout << ",\"override_a_after_reassign\":";
  writeEffects(std::cout,overrideA,architecture);
  std::cout << ",\"lookup_record_before_first\":{\"overlap\":"
            << beforeFirstOverlap << ",\"disjoint\":" << beforeFirstDisjoint << '}';
  std::cout << ",\"set_model\":{\"initial\":{\"extrapop\":"
            << initialExtraPop << ",\"has_this\":" << initialThis
            << ",\"constructor\":" << initialConstructor
            << ",\"auto_killed\":" << initialAutoKilled
            << "},\"unknown_switch\":{\"extrapop\":" << switchedExtraPop
            << ",\"has_this\":" << switchedThis
            << ",\"constructor\":" << switchedConstructor
            << ",\"auto_killed\":" << switchedAutoKilled
            << "},\"known_switch\":{\"extrapop\":" << replacedExtraPop
            << ",\"has_this\":" << replacedThis
            << ",\"constructor\":" << replacedConstructor
            << ",\"auto_killed\":" << replacedAutoKilled
            << "},\"null_switch\":{\"has_model\":" << sticky.hasModel()
            << ",\"extrapop\":" << sticky.getExtraPop()
            << ",\"has_this\":" << sticky.hasThisPointer()
            << ",\"constructor\":" << sticky.isConstructor()
            << ",\"auto_killed\":" << sticky.isAutoKilledByCall()
            << ",\"print_has_no_model\":" << nullPrintHasNoModel
            << "}},\"output_lock_auto_killed\":{\"before\":" << outputLockBefore
            << ",\"after\":" << outputLockAfter
            << "},\"copy_preserves_destination_return_bytes\":"
            << copyHintDestination.getReturnBytesConsumed() << "}\n";
}

} // namespace

int main(int argc,char **argv)
{
  try {
    if (argc != 3)
      throw std::invalid_argument("usage: funcproto_effect_model_1204 SPEC_DIRECTORY BINARY");
    runFixture(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &err) {
    std::cerr << "LowlevelError: " << err.explain << '\n';
  }
  catch(const std::exception &err) {
    std::cerr << "std::exception: " << err.what() << '\n';
  }
  return 1;
}

File: tests/oracle/funcproto_effect_model_1204.rs
use std::sync::Arc;

use rugra::fspec::{EffectRecord, EffectType, FuncProto, ProtoModelFull};
use rugra::space::AddressSpace;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeMetatype};

#[derive(Clone, Copy)]
struct Probe {
    label: &'static str,
    space_name: &'static str,
    space: AddressSpace,
    offset: u64,
    size: i32,
}

const PROBES: &[Probe] = &[
    Probe { label: "unique_always", space_name: "unique", space: AddressSpace::Unique, offset: 0x777, size: 4 },
    Probe { label: "same_offset_const", space_name: "const", space: AddressSpace::Const, offset: 0x10, size: 8 },
    Probe { label: "whole_ram_space", space_name: "ram", space: AddressSpace::Ram, offset: 0x777, size: 32 },
    Probe { label: "register_exact", space_name: "register", space: AddressSpace::Register, offset: 0x10, size: 8 },
    Probe { label: "register_contained", space_name: "register", space: AddressSpace::Register, offset: 0x12, size: 2 },
    Probe { label: "register_partial_left", space_name: "register", space: AddressSpace::Register, offset: 0x0e, size: 4 },
    Probe { label: "register_partial_right", space_name: "register", space: AddressSpace::Register, offset: 0x16, size: 4 },
    Probe { label: "register_killed", space_name: "register", space: AddressSpace::Register, offset: 0x20, size: 8 },
    Probe { label: "return_address", space_name: "register", space: AddressSpace::Register, offset: 0x30, size: 8 },
    Probe { label: "same_offset_stack", space_name: "stack", space: AddressSpace::Stack, offset: 0x10, size: 8 },
    Probe { label: "same_offset_register", space_name: "register", space: AddressSpace::Register, offset: 0x10, size: 8 },
    Probe { label: "before_first_register", space_name: "register", space: AddressSpace::Register, offset: 0x08, size: 8 },
];

fn effect_name(effect: EffectType) -> &'static str {
    match effect {
        EffectType::Unaffected => "unaffected",
        EffectType::KilledByCall => "killedbycall",
        EffectType::ReturnAddress => "return_address",
        EffectType::UnknownEffect => "unknown_effect",
    }
}

fn model_effects() -> Vec<EffectRecord> {
    let mut effects = vec![
        EffectRecord::new(AddressSpace::Const, 0x10, 8, EffectType::KilledByCall),
        EffectRecord::new(AddressSpace::Ram, 0, 0, EffectType::Unaffected),
        EffectRecord::new(AddressSpace::Register, 0x10, 8, EffectType::Unaffected),
        EffectRecord::new(AddressSpace::Register, 0x20, 8, EffectType::KilledByCall),
        EffectRecord::new(AddressSpace::Register, 0x30, 8, EffectType::ReturnAddress),
        EffectRecord::new(AddressSpace::Stack, 0x10, 8, EffectType::KilledByCall),
    ];
    effects.sort_by_key(|effect| (effect.space.space_id(), effect.offset));
    effects
}

fn space_name(space: AddressSpace) -> &'static str {
    match space {
        AddressSpace::Const => "const",
        AddressSpace::Unique => "unique",
        AddressSpace::Ram => "ram",
        AddressSpace::Register => "register",
        AddressSpace::Stack => "stack",
        _ => panic!("fixture received an unexpected address space"),
    }
}

fn write_effective_records(out: &mut String, proto: &FuncProto) {
    out.push('[');
    for (index, effect) in proto.effect_iter().iter().enumerate() {
        if index != 0 {
            out.push(',');
        }
        out.push_str("{\"space\":\"");
        out.push_str(space_name(effect.space));
        out.push_str("\",\"offset\":");
        out.push_str(&effect.offset.to_string());
        out.push_str(",\"size\":");
        out.push_str(&effect.size.to_string());
        out.push_str(",\"effect\":\"");
        out.push_str(effect_name(effect.effect_type));
        out.push_str("\"}");
    }
    out.push(']');
}

fn make_model(name: &str, extrapop: i32, has_this: bool, constructor: bool) -> Arc<ProtoModelFull> {
    let mut model = ProtoModelFull::new(Some(AddressSpace::Stack), 8);
    model.name = name.to_string();
    model.extrapop = extrapop;
    model.has_this = has_this;
    model.is_construct = constructor;
    model.output.set_auto_killed_by_call(true);
    model.effectlist = model_effects();
    Arc::new(model)
}

fn make_unknown_model() -> Arc<ProtoModelFull> {
    let mut model = ProtoModelFull::new(Some(AddressSpace::Stack), 8);
    model.name = "fixture_unknown".to_string();
    model.extrapop = 0x8000;
    model.output.set_auto_killed_by_call(true);
    model.effectlist = vec![EffectRecord::new(
        AddressSpace::Register,
        0x40,
        8,
        EffectType::ReturnAddress,
    )];
    Arc::new(model)
}

fn make_replacement_model() -> Arc<ProtoModelFull> {
    let mut model = ProtoModelFull::new(Some(AddressSpace::Stack), 8);
    model.name = "fixture_replacement".to_string();
    model.extrapop = 24;
    model.output.set_auto_killed_by_call(true);
    Arc::new(model)
}

fn void_type() -> Arc<Datatype> {
    Arc::new(Datatype::Void(TypeBase::new(
        "void".to_string(),
        0,
        TypeMetatype::Void,
    )))
}

fn write_effects(out: &mut String, proto: &FuncProto) {
    out.push('[');
    for (index, probe) in PROBES.iter().enumerate() {
        if index != 0 {
            out.push(',');
        }
        let effect = proto.has_effect(probe.space, probe.offset, probe.size);
        out.push_str("{\"label\":\"");
        out.push_str(probe.label);
        out.push_str("\",\"space\":\"");
        out.push_str(probe.space_name);
        out.push_str("\",\"offset\":");
        out.push_str(&probe.offset.to_string());
        out.push_str(",\"size\":");
        out.push_str(&probe.size.to_string());
        out.push_str(",\"effect\":\"");
        out.push_str(effect_name(effect));
        out.push_str("\"}");
    }
    out.push(']');
}

fn lookup_record_code(result: Result<Option<usize>, ()>) -> i32 {
    match result {
        Ok(Some(index)) => index as i32,
        Ok(None) => -1,
        Err(()) => -2,
    }
}

fn main() {
    let void_type = void_type();
    let model = make_model("fixture_model", 16, true, true);
    let independent_model = make_model("fixture_model", 16, true, true);
    let unknown_model = make_unknown_model();
    let replacement_model = make_replacement_model();

    let mut model_proto = FuncProto::new(String::new(), void_type.clone());
    model_proto.set_model(Some(model.clone()));
    let mut model_copy = FuncProto::new(String::new(), void_type.clone());
    model_copy.copy_from(&model_proto);
    let mut independent_proto = FuncProto::new(String::new(), void_type.clone());
    independent_proto.set_model(Some(independent_model));

    let mut override_a = FuncProto::new(String::new(), void_type.clone());
    override_a.set_model(Some(model.clone()));
    override_a.effects = vec![
        EffectRecord::new(AddressSpace::Register, 0x10, 8, EffectType::KilledByCall),
    ];
    override_a
        .decode_effect(&model.effectlist)
        .expect("exact override A must merge");
    let mut override_a_copy = FuncProto::new(String::new(), void_type.clone());
    override_a_copy.copy_from(&override_a);

    let mut override_b = FuncProto::new(String::new(), void_type.clone());
    override_b.set_model(Some(model.clone()));
    override_b.effects = vec![
        EffectRecord::new(AddressSpace::Register, 0x30, 8, EffectType::Unaffected),
    ];
    override_b
        .decode_effect(&model.effectlist)
        .expect("exact override B must merge");
    override_a.copy_from(&override_b);

    let lookup_records = vec![EffectRecord::new(
        AddressSpace::Register,
        0x10,
        8,
        EffectType::Unaffected,
    )];
    let before_first_overlap = lookup_record_code(ProtoModelFull::lookup_record(
        &lookup_records,
        1,
        AddressSpace::Register,
        0x08,
        16,
    ));
    let before_first_disjoint = lookup_record_code(ProtoModelFull::lookup_record(
        &lookup_records,
        1,
        AddressSpace::Register,
        0x00,
        8,
    ));

    let mut sticky = FuncProto::new(String::new(), void_type.clone());
    sticky.set_model(Some(model));
    let initial = (
        sticky.get_extra_pop(),
        sticky.has_thisptr(),
        sticky.is_constructor_flag(),
        sticky.is_auto_killed_by_call(),
    );
    sticky.set_model(Some(unknown_model.clone()));
    let switched = (
        sticky.get_extra_pop(),
        sticky.has_thisptr(),
        sticky.is_constructor_flag(),
        sticky.is_auto_killed_by_call(),
    );
    sticky.set_model(Some(replacement_model));
    let replaced = (
        sticky.get_extra_pop(),
        sticky.has_thisptr(),
        sticky.is_constructor_flag(),
        sticky.is_auto_killed_by_call(),
    );
    sticky.set_model(None);
    let mut null_model_print = String::new();
    sticky.print_raw("fixture", &mut null_model_print);
    let null_print_has_no_model = null_model_print.starts_with("(no model) ");

    let mut output_locked = FuncProto::new(String::new(), void_type.clone());
    let output_lock_before = output_locked.is_auto_killed_by_call();
    output_locked.set_output_lock(true);
    let output_lock_after = output_locked.is_auto_killed_by_call();

    let mut copy_hint_source = FuncProto::new(String::new(), void_type.clone());
    copy_hint_source.set_return_bytes_consumed(7);
    let mut copy_hint_destination = FuncProto::new(String::new(), void_type);
    copy_hint_destination.set_return_bytes_consumed(3);
    copy_hint_destination.copy_from(&copy_hint_source);

    let mut out = String::from(
        "{\"schema\":1,\"fixture\":\"PROTO-EFFECT-MODEL-0001\",\"model_effects\":",
    );
    write_effects(&mut out, &model_proto);
    out.push_str(",\"effective_model_records\":");
    write_effective_records(&mut out, &model_proto);
    out.push_str(",\"model_identity\":{\"copy\":");
    out.push(if model_proto.shares_model_with(&model_copy) { '1' } else { '0' });
    out.push_str(",\"independent_same_definition\":");
    out.push(if model_proto.shares_model_with(&independent_proto) { '1' } else { '0' });
    out.push_str("},\"override_a_copy\":");
    write_effects(&mut out, &override_a_copy);
    out.push_str(",\"effective_override_records\":");
    write_effective_records(&mut out, &override_a_copy);
    out.push_str(",\"override_a_after_reassign\":");
    write_effects(&mut out, &override_a);
    out.push_str(",\"lookup_record_before_first\":{\"overlap\":");
    out.push_str(&before_first_overlap.to_string());
    out.push_str(",\"disjoint\":");
    out.push_str(&before_first_disjoint.to_string());
    out.push('}');
    out.push_str(",\"set_model\":{\"initial\":{\"extrapop\":");
    out.push_str(&initial.0.to_string());
    out.push_str(",\"has_this\":");
    out.push(if initial.1 { '1' } else { '0' });
    out.push_str(",\"constructor\":");
    out.push(if initial.2 { '1' } else { '0' });
    out.push_str(",\"auto_killed\":");
    out.push(if initial.3 { '1' } else { '0' });
    out.push_str("},\"unknown_switch\":{\"extrapop\":");
    out.push_str(&switched.0.to_string());
    out.push_str(",\"has_this\":");
    out.push(if switched.1 { '1' } else { '0' });
    out.push_str(",\"constructor\":");
    out.push(if switched.2 { '1' } else { '0' });
    out.push_str(",\"auto_killed\":");
    out.push(if switched.3 { '1' } else { '0' });
    out.push_str("},\"known_switch\":{\"extrapop\":");
    out.push_str(&replaced.0.to_string());
    out.push_str(",\"has_this\":");
    out.push(if replaced.1 { '1' } else { '0' });
    out.push_str(",\"constructor\":");
    out.push(if replaced.2 { '1' } else { '0' });
    out.push_str(",\"auto_killed\":");
    out.push(if replaced.3 { '1' } else { '0' });
    out.push_str("},\"null_switch\":{\"has_model\":");
    out.push(if sticky.has_model() { '1' } else { '0' });
    out.push_str(",\"extrapop\":");
    out.push_str(&sticky.get_extra_pop().to_string());
    out.push_str(",\"has_this\":");
    out.push(if sticky.has_thisptr() { '1' } else { '0' });
    out.push_str(",\"constructor\":");
    out.push(if sticky.is_constructor_flag() { '1' } else { '0' });
    out.push_str(",\"auto_killed\":");
    out.push(if sticky.is_auto_killed_by_call() { '1' } else { '0' });
    out.push_str(",\"print_has_no_model\":");
    out.push(if null_print_has_no_model { '1' } else { '0' });
    out.push_str("}},\"output_lock_auto_killed\":{\"before\":");
    out.push(if output_lock_before { '1' } else { '0' });
    out.push_str(",\"after\":");
    out.push(if output_lock_after { '1' } else { '0' });
    out.push_str("},\"copy_preserves_destination_return_bytes\":");
    out.push_str(&copy_hint_destination.get_return_bytes_consumed().to_string());
    out.push_str("}\n");
    print!("{out}");
}
